internal/pkg/vm/internal/vmcontext/testing.go

```diff
@@ -326,6 +322,55 @@ func (a *ValidationApplier) ApplyMessage(context *vtypes.ExecutionContext, state
 	return receipt, penalty, reward, nil
 }
 
+func (a *ValidationApplier) ApplySignedMessage(context *vtypes.ExecutionContext, state vstate.VMWrapper, msg *vtypes.SignedMessage) (vtypes.MessageReceipt, abi.TokenAmount, abi.TokenAmount, error) {
+	st := state.(*ValidationVMWrapper)
+
+	// set epoch
+	// Note: this would have normally happened during `ApplyTipset()`
+	st.vm.currentEpoch = context.Epoch
+	st.vm.pricelist = gascost.PricelistByEpoch(context.Epoch)
+
+	// map message
+	// Dragons: fix after cleaning up our msg
+	ourmsg := &types.SignedMessage{
+		Message:   *toOurMessage(&msg.Message),
+		Signature: msg.Signature,
+	}
+
+	// invoke vm
+	ourreceipt, penalty, reward := st.vm.applyMessage(&ourmsg.Message, ourmsg.OnChainLen(), &fakeRandSrc{})
+
+	// commit and persist changes
+	// Note: this is not done on production for each msg
+	if err := st.PersistChanges(); err != nil {
+		return vtypes.MessageReceipt{}, big.Zero(), big.Zero(), err
+	}
+
+	// map receipt
+	receipt := vtypes.MessageReceipt{
+		ExitCode:    ourreceipt.ExitCode,
+		ReturnValue: ourreceipt.ReturnValue,
+		GasUsed:     vtypes.GasUnits(ourreceipt.GasUsed),
+	}
+
+	return receipt, penalty, reward, nil
```

Comment:
Please factor out aaaaallll the repeated code to a `applyMessageInner()` or similar.